File: sorting/quick_sort.go
// Quicksort
/*
	In this implementation, we use the standard algorithm for quicksort:

	1. Choose a pivot element (we choose the middle element in this case)
	2. Partition the array into two subarrays: elements less than the pivot go to the left subarray,
	   and elements greater than the pivot go to the right subarray
	3. Recursively sort the left and right subarrays using the same algorithm

	The quicksort function takes three arguments: the array to be sorted, the left index of the subarray,
	and the right index of the subarray. We start by choosing the pivot element (in this case, the middle element),
	initializing two pointers to the left and right ends of the subarray, and then partitioning the array into two
	subarrays by moving elements less than the pivot to the left subarray and elements greater than the pivot to
	the right subarray. We repeat this process until the subarrays have length 0 or 1.

	Once the subarrays are partitioned, we recursively sort the left and right subarrays using the quicksort function.
	We use two conditional statements to check if the left or right subarray has length greater than 1 and call quicksort
	on that subarray if it does. Finally, the sorted array is returned.
*/
package main

import "fmt"

// Quicksort function to sort an array of integers in ascending order
func quicksort(arr []int, left, right int) {
	// Choose pivot element (middle element in this case)
	pivot := arr[(left+right)/2]

	// Initialize two pointers to left and right ends of subarray
	i := left
	j := right

	// Partition array into two subarrays: elements less than pivot
	// go to the left subarray and elements greater than pivot go to
	// the right subarray. Repeat until subarrays have length 0 or 1.
	for i <= j {
		// Find first element in left subarray greater than pivot
		for arr[i] < pivot {
			i++
		}
		// Find first element in right subarray less than pivot
		for arr[j] > pivot {
			j--
		}
		// Swap elements at i and j if they are in the wrong subarray
		if i <= j {
			arr[i], arr[j] = arr[j], arr[i]
			i++
			j--
		}
	}

	// Recursively sort left and right subarrays
	if left < j {
		quicksort(arr, left, j)
	}
	if i < right {
		quicksort(arr, i, right)
	}
}

func main() {
	// Example usage
	arr := []int{3, 7, 1, 8, 4, 2, 9, 5, 6}
	quicksort(arr, 0, len(arr)-1)
	fmt.Println(arr) // Output: [1 2 3 4 5 6 7 8 9]
}

File: sorting/quick_sort.js
/*
    The quickSort function implements the quicksort algorithm, which takes an array arr and the indices low and high 
    of the subarray to be sorted. If low is less than high, the array is partitioned around a pivot element using the 
    partition function, and the subarrays on either side of the pivot are recursively sorted using the quickSort function.

    The partition function takes an array arr and the indices low and high of the subarray to be partitioned. 
    It chooses the rightmost element as the pivot, initializes the index of the smaller element to low - 1, and 
    iterates over the subarray. If an element arr[j] is less than the pivot, it is swapped with the element arr[i+1] 
    and the index i is incremented. Finally, the pivot element is swapped with the element arr[i+1], 
    and the index i+1 is returned as the index of the pivot.

    The time complexity of quicksort is O(n log n) on average, and O(n^2) in the worst case. The space complexity 
    is O(log n) on average, and O(n) in the worst case.
*/
/**
 * This function implements the quicksort algorithm to sort an array in place.
 * @param {Array} arr - The array to be sorted.
 * @param {number} low - The lower index of the subarray to be sorted.
 * @param {number} high - The higher index of the subarray to be sorted.
 * @returns {void}
 */
function quickSort(arr, low, high) {
  if (low < high) {
    // Partition the array around the pivot element and get the index of the pivot.
    const pivotIndex = partition(arr, low, high);

    // Recursively sort the subarrays on either side of the pivot.
    quickSort(arr, low, pivotIndex - 1);
    quickSort(arr, pivotIndex + 1, high);
  }
}

/**
 * This function partitions an array around a pivot element and returns the index of the pivot.
 * @param {Array} arr - The array to be partitioned.
 * @param {number} low - The lower index of the subarray to be partitioned.
 * @param {number} high - The higher index of the subarray to be partitioned.
 * @returns {number} - The index of the pivot element.
 */
function partition(arr, low, high) {
  // Choose the rightmost element as the pivot.
  const pivot = arr[high];

  // Initialize the index of the smaller element.
  let i = low - 1;

  // Iterate over the subarray and partition it.
  for (let j = low; j < high; j++) {
    if (arr[j] < pivot) {
      i++;

      // Swap arr[i] and arr[j].
      const temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
    }
  }

  // Swap arr[i+1] and arr[high] (or the pivot).
  const temp = arr[i + 1];
  arr[i + 1] = arr[high];
  arr[high] = temp;

  // Return the index of the pivot.
  return i + 1;
}
